File: client/src/constants/permissions.ts
export const Permissions = {
  Customer: {
    ViewProfile: 'Customer.ViewProfile',
    EditProfile: 'Customer.EditProfile',
    ViewList: 'Customer.ViewList',
    Delete: 'Customer.Delete',
  },
  Vendor: {
    ViewProfile: 'Vendor.ViewProfile',
    EditProfile: 'Vendor.EditProfile',
    ViewDashboard: 'Vendor.ViewDashboard',
    ViewList: 'Vendor.ViewList',
    Create: 'Vendor.Create',
    Delete: 'Vendor.Delete',
  },
  Admin: {
    ViewProfile: 'Admin.ViewProfile',
    EditProfile: 'Admin.EditProfile',
    ViewList: 'Admin.ViewList',
    Create: 'Admin.Create',
    Delete: 'Admin.Delete',
    ManageUsers: 'Admin.ManageUsers',
    ManageRoles: 'Admin.ManageRoles',
    ManageClaims: 'Admin.ManageClaims',
  },
  Auth: {
    ChangePassword: 'Auth.ChangePassword',
    ViewOwnProfile: 'Auth.ViewOwnProfile',
    EditOwnProfile: 'Auth.EditOwnProfile',
  },
  Product: {
    ViewList: 'Product.ViewList',
    View: 'Product.View',
    Create: 'Product.Create',
    Edit: 'Product.Edit',
    Delete: 'Product.Delete',
  },
  Category: {
    ViewList: 'Category.ViewList',
    View: 'Category.View',
    Create: 'Category.Create',
    Edit: 'Category.Edit',
    Delete: 'Category.Delete',
  },
  SubCategory: {
    ViewList: 'SubCategory.ViewList',
    View: 'SubCategory.View',
    Create: 'SubCategory.Create',
    Edit: 'SubCategory.Edit',
    Delete: 'SubCategory.Delete',
  },
  Brand: {
    ViewList: 'Brand.ViewList',
    View: 'Brand.View',
    Create: 'Brand.Create',
    Edit: 'Brand.Edit',
    Delete: 'Brand.Delete',
  },
  UnitOfMeasure: {
    ViewList: 'UnitOfMeasure.ViewList',
    View: 'UnitOfMeasure.View',
    Create: 'UnitOfMeasure.Create',
    Edit: 'UnitOfMeasure.Edit',
    Delete: 'UnitOfMeasure.Delete',
  },
  Warranty: {
    ViewList: 'Warranty.ViewList',
    View: 'Warranty.View',
    Create: 'Warranty.Create',
    Edit: 'Warranty.Edit',
    Delete: 'Warranty.Delete',
  },
  VariantAttribute: {
    ViewList: 'VariantAttribute.ViewList',
    View: 'VariantAttribute.View',
    Create: 'VariantAttribute.Create',
    Edit: 'VariantAttribute.Edit',
    Delete: 'VariantAttribute.Delete',
  },
  Account: {
    ViewList: 'Account.ViewList',
    View: 'Account.View',
    Create: 'Account.Create',
    Edit: 'Account.Edit',
    Delete: 'Account.Delete',
  },
  Coupon: {
    ViewList: 'Coupon.ViewList',
    View: 'Coupon.View',
    Create: 'Coupon.Create',
    Edit: 'Coupon.Edit',
    Delete: 'Coupon.Delete',
  },
  Discount: {
    ViewList: 'Discount.ViewList',
    View: 'Discount.View',
    Create: 'Discount.Create',
    Edit: 'Discount.Edit',
    Delete: 'Discount.Delete',
  },
  GiftCard: {
    ViewList: 'GiftCard.ViewList',
    View: 'GiftCard.View',
    Create: 'GiftCard.Create',
    Edit: 'GiftCard.Edit',
    Delete: 'GiftCard.Delete',
  },
} as const;

export type Permission = typeof Permissions[keyof typeof Permissions][keyof typeof Permissions[keyof typeof Permissions]];

export const PermissionGroups = [
  {
    name: 'Customer',
    label: 'Customer Permissions',
    permissions: [
      { key: Permissions.Customer.ViewProfile, label: 'View Profile' },
      { key: Permissions.Customer.EditProfile, label: 'Edit Profile' },
      { key: Permissions.Customer.ViewList, label: 'View List' },
      { key: Permissions.Customer.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Vendor',
    label: 'Vendor Permissions',
    permissions: [
      { key: Permissions.Vendor.ViewProfile, label: 'View Profile' },
      { key: Permissions.Vendor.EditProfile, label: 'Edit Profile' },
      { key: Permissions.Vendor.ViewDashboard, label: 'View Dashboard' },
      { key: Permissions.Vendor.ViewList, label: 'View List' },
      { key: Permissions.Vendor.Create, label: 'Create' },
      { key: Permissions.Vendor.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Admin',
    label: 'Admin Permissions',
    permissions: [
      { key: Permissions.Admin.ViewProfile, label: 'View Profile' },
      { key: Permissions.Admin.EditProfile, label: 'Edit Profile' },
      { key: Permissions.Admin.ViewList, label: 'View List' },
      { key: Permissions.Admin.Create, label: 'Create' },
      { key: Permissions.Admin.Delete, label: 'Delete' },
      { key: Permissions.Admin.ManageUsers, label: 'Manage Users' },
      { key: Permissions.Admin.ManageRoles, label: 'Manage Roles' },
      { key: Permissions.Admin.ManageClaims, label: 'Manage Claims' },
    ],
  },
  {
    name: 'Auth',
    label: 'Authentication Permissions',
    permissions: [
      { key: Permissions.Auth.ChangePassword, label: 'Change Password' },
      { key: Permissions.Auth.ViewOwnProfile, label: 'View Own Profile' },
      { key: Permissions.Auth.EditOwnProfile, label: 'Edit Own Profile' },
    ],
  },
  {
    name: 'Product',
    label: 'Product Permissions',
    permissions: [
      { key: Permissions.Product.ViewList, label: 'View List' },
      { key: Permissions.Product.View, label: 'View' },
      { key: Permissions.Product.Create, label: 'Create' },
      { key: Permissions.Product.Edit, label: 'Edit' },
      { key: Permissions.Product.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Category',
    label: 'Category Permissions',
    permissions: [
      { key: Permissions.Category.ViewList, label: 'View List' },
      { key: Permissions.Category.View, label: 'View' },
      { key: Permissions.Category.Create, label: 'Create' },
      { key: Permissions.Category.Edit, label: 'Edit' },
      { key: Permissions.Category.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'SubCategory',
    label: 'SubCategory Permissions',
    permissions: [
      { key: Permissions.SubCategory.ViewList, label: 'View List' },
      { key: Permissions.SubCategory.View, label: 'View' },
      { key: Permissions.SubCategory.Create, label: 'Create' },
      { key: Permissions.SubCategory.Edit, label: 'Edit' },
      { key: Permissions.SubCategory.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Brand',
    label: 'Brand Permissions',
    permissions: [
      { key: Permissions.Brand.ViewList, label: 'View List' },
      { key: Permissions.Brand.View, label: 'View' },
      { key: Permissions.Brand.Create, label: 'Create' },
      { key: Permissions.Brand.Edit, label: 'Edit' },
      { key: Permissions.Brand.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'UnitOfMeasure',
    label: 'Unit Of Measure Permissions',
    permissions: [
      { key: Permissions.UnitOfMeasure.ViewList, label: 'View List' },
      { key: Permissions.UnitOfMeasure.View, label: 'View' },
      { key: Permissions.UnitOfMeasure.Create, label: 'Create' },
      { key: Permissions.UnitOfMeasure.Edit, label: 'Edit' },
      { key: Permissions.UnitOfMeasure.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Warranty',
    label: 'Warranty Permissions',
    permissions: [
      { key: Permissions.Warranty.ViewList, label: 'View List' },
      { key: Permissions.Warranty.View, label: 'View' },
      { key: Permissions.Warranty.Create, label: 'Create' },
      { key: Permissions.Warranty.Edit, label: 'Edit' },
      { key: Permissions.Warranty.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'VariantAttribute',
    label: 'Variant Attribute Permissions',
    permissions: [
      { key: Permissions.VariantAttribute.ViewList, label: 'View List' },
      { key: Permissions.VariantAttribute.View, label: 'View' },
      { key: Permissions.VariantAttribute.Create, label: 'Create' },
      { key: Permissions.VariantAttribute.Edit, label: 'Edit' },
      { key: Permissions.VariantAttribute.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Account',
    label: 'Account Permissions',
    permissions: [
      { key: Permissions.Account.ViewList, label: 'View List' },
      { key: Permissions.Account.View, label: 'View' },
      { key: Permissions.Account.Create, label: 'Create' },
      { key: Permissions.Account.Edit, label: 'Edit' },
      { key: Permissions.Account.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Coupon',
    label: 'Coupon Permissions',
    permissions: [
      { key: Permissions.Coupon.ViewList, label: 'View List' },
      { key: Permissions.Coupon.View, label: 'View' },
      { key: Permissions.Coupon.Create, label: 'Create' },
      { key: Permissions.Coupon.Edit, label: 'Edit' },
      { key: Permissions.Coupon.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'Discount',
    label: 'Discount Permissions',
    permissions: [
      { key: Permissions.Discount.ViewList, label: 'View List' },
      { key: Permissions.Discount.View, label: 'View' },
      { key: Permissions.Discount.Create, label: 'Create' },
      { key: Permissions.Discount.Edit, label: 'Edit' },
      { key: Permissions.Discount.Delete, label: 'Delete' },
    ],
  },
  {
    name: 'GiftCard',
    label: 'Gift Card Permissions',
    permissions: [
      { key: Permissions.GiftCard.ViewList, label: 'View List' },
      { key: Permissions.GiftCard.View, label: 'View' },
      { key: Permissions.GiftCard.Create, label: 'Create' },
      { key: Permissions.GiftCard.Edit, label: 'Edit' },
      { key: Permissions.GiftCard.Delete, label: 'Delete' },
    ],
  },
];

export function getAllPermissions(): string[] {
  return Object.values(Permissions).flatMap((group) => Object.values(group));
}

export function getPermissionsForRole(role: string): string[] {
  switch (role) {
    case 'Customer':
      return [
        Permissions.Customer.ViewProfile,
        Permissions.Customer.EditProfile,
        Permissions.Auth.ChangePassword,
        Permissions.Auth.ViewOwnProfile,
        Permissions.Auth.EditOwnProfile,
        Permissions.Product.ViewList,
        Permissions.Product.View,
      ];
    case 'Vendor':
      return [
        Permissions.Vendor.ViewProfile,
        Permissions.Vendor.EditProfile,
        Permissions.Vendor.ViewDashboard,
        Permissions.Auth.ChangePassword,
        Permissions.Auth.ViewOwnProfile,
        Permissions.Auth.EditOwnProfile,
        Permissions.Product.ViewList,
        Permissions.Product.View,
        Permissions.Product.Create,
        Permissions.Product.Edit,
      ];
    case 'Admin':
    case 'SuperAdmin':
      return getAllPermissions();
    default:
      return [];
  }
}

File: Infrastructure/Data/Authorization/Permissions.cs
namespace Infrastructure.Data.Authorization;

public static class Permissions
{
    public static class Customer
    {
        public const string ViewProfile = "Customer.ViewProfile";
        public const string EditProfile = "Customer.EditProfile";
        public const string ViewList = "Customer.ViewList";
        public const string Delete = "Customer.Delete";
    }

    public static class Vendor
    {
        public const string ViewProfile = "Vendor.ViewProfile";
        public const string EditProfile = "Vendor.EditProfile";
        public const string ViewDashboard = "Vendor.ViewDashboard";
        public const string ViewList = "Vendor.ViewList";
        public const string Create = "Vendor.Create";
        public const string Delete = "Vendor.Delete";
    }

    public static class Admin
    {
        public const string ViewProfile = "Admin.ViewProfile";
        public const string EditProfile = "Admin.EditProfile";
        public const string ViewList = "Admin.ViewList";
        public const string Create = "Admin.Create";
        public const string Delete = "Admin.Delete";
        public const string ManageUsers = "Admin.ManageUsers";
        public const string ManageRoles = "Admin.ManageRoles";
        public const string ManageClaims = "Admin.ManageClaims";
    }

    public static class Auth
    {
        public const string ChangePassword = "Auth.ChangePassword";
        public const string ViewOwnProfile = "Auth.ViewOwnProfile";
        public const string EditOwnProfile = "Auth.EditOwnProfile";
    }

    public static class Product
    {
        public const string ViewList = "Product.ViewList";
        public const string View = "Product.View";
        public const string Create = "Product.Create";
        public const string Edit = "Product.Edit";
        public const string Delete = "Product.Delete";
    }

    public static class Category
    {
        public const string ViewList = "Category.ViewList";
        public const string View = "Category.View";
        public const string Create = "Category.Create";
        public const string Edit = "Category.Edit";
        public const string Delete = "Category.Delete";
    }

    public static class SubCategory
    {
        public const string ViewList = "SubCategory.ViewList";
        public const string View = "SubCategory.View";
        public const string Create = "SubCategory.Create";
        public const string Edit = "SubCategory.Edit";
        public const string Delete = "SubCategory.Delete";
    }

    public static class Brand
    {
        public const string ViewList = "Brand.ViewList";
        public const string View = "Brand.View";
        public const string Create = "Brand.Create";
        public const string Edit = "Brand.Edit";
        public const string Delete = "Brand.Delete";
    }

    public static class UnitOfMeasure
    {
        public const string ViewList = "UnitOfMeasure.ViewList";
        public const string View = "UnitOfMeasure.View";
        public const string Create = "UnitOfMeasure.Create";
        public const string Edit = "UnitOfMeasure.Edit";
        public const string Delete = "UnitOfMeasure.Delete";
    }

    public static class Warranty
    {
        public const string ViewList = "Warranty.ViewList";
        public const string View = "Warranty.View";
        public const string Create = "Warranty.Create";
        public const string Edit = "Warranty.Edit";
        public const string Delete = "Warranty.Delete";
    }

    public static class VariantAttribute
    {
        public const string ViewList = "VariantAttribute.ViewList";
        public const string View = "VariantAttribute.View";
        public const string Create = "VariantAttribute.Create";
        public const string Edit = "VariantAttribute.Edit";
        public const string Delete = "VariantAttribute.Delete";
    }

    public static class Account
    {
        public const string ViewList = "Account.ViewList";
        public const string View = "Account.View";
        public const string Create = "Account.Create";
        public const string Edit = "Account.Edit";
        public const string Delete = "Account.Delete";
    }

    public static class Coupon
    {
        public const string ViewList = "Coupon.ViewList";
        public const string View = "Coupon.View";
        public const string Create = "Coupon.Create";
        public const string Edit = "Coupon.Edit";
        public const string Delete = "Coupon.Delete";
    }

    public static class Discount
    {
        public const string ViewList = "Discount.ViewList";
        public const string View = "Discount.View";
        public const string Create = "Discount.Create";
        public const string Edit = "Discount.Edit";
        public const string Delete = "Discount.Delete";
    }

    public static class GiftCard
    {
        public const string ViewList = "GiftCard.ViewList";
        public const string View = "GiftCard.View";
        public const string Create = "GiftCard.Create";
        public const string Edit = "GiftCard.Edit";
        public const string Delete = "GiftCard.Delete";
    }

    public static List<string> GetAll()
    {
        return new List<string>
        {
            Customer.ViewProfile,
            Customer.EditProfile,
            Customer.ViewList,
            Customer.Delete,
            Vendor.ViewProfile,
            Vendor.EditProfile,
            Vendor.ViewDashboard,
            Vendor.ViewList,
            Vendor.Create,
            Vendor.Delete,
            Admin.ViewProfile,
            Admin.EditProfile,
            Admin.ViewList,
            Admin.Create,
            Admin.Delete,
            Admin.ManageUsers,
            Admin.ManageRoles,
            Admin.ManageClaims,
            Auth.ChangePassword,
            Auth.ViewOwnProfile,
            Auth.EditOwnProfile,
            Product.ViewList,
            Product.View,
            Product.Create,
            Product.Edit,
            Product.Delete,
            Category.ViewList,
            Category.View,
            Category.Create,
            Category.Edit,
            Category.Delete,
            SubCategory.ViewList,
            SubCategory.View,
            SubCategory.Create,
            SubCategory.Edit,
            SubCategory.Delete,
            Brand.ViewList,
            Brand.View,
            Brand.Create,
            Brand.Edit,
            Brand.Delete,
            UnitOfMeasure.ViewList,
            UnitOfMeasure.View,
            UnitOfMeasure.Create,
            UnitOfMeasure.Edit,
            UnitOfMeasure.Delete,
            Warranty.ViewList,
            Warranty.View,
            Warranty.Create,
            Warranty.Edit,
            Warranty.Delete,
            VariantAttribute.ViewList,
            VariantAttribute.View,
            VariantAttribute.Create,
            VariantAttribute.Edit,
            VariantAttribute.Delete,
            Account.ViewList,
            Account.View,
            Account.Create,
            Account.Edit,
            Account.Delete,
            Coupon.ViewList,
            Coupon.View,
            Coupon.Create,
            Coupon.Edit,
            Coupon.Delete,
            Discount.ViewList,
            Discount.View,
            Discount.Create,
            Discount.Edit,
            Discount.Delete,
            GiftCard.ViewList,
            GiftCard.View,
            GiftCard.Create,
            GiftCard.Edit,
            GiftCard.Delete
        };
    }

    public static List<string> GetDefaultForRole(string roleName)
    {
        return roleName switch
        {
            Roles.Customer => new List<string>
            {
                Customer.ViewProfile,
                Customer.EditProfile,
                Auth.ChangePassword,
                Auth.ViewOwnProfile,
                Auth.EditOwnProfile,
                Product.ViewList,
                Product.View
            },
            Roles.Merchant => new List<string>
            {
                Vendor.ViewProfile,
                Vendor.EditProfile,
                Vendor.ViewDashboard,
                Auth.ChangePassword,
                Auth.ViewOwnProfile,
                Auth.EditOwnProfile,
                Product.ViewList,
                Product.View,
                Product.Create,
                Product.Edit
            },
            Roles.StaffMerchant => new List<string>
            {
                Vendor.ViewProfile,
                Vendor.EditProfile,
                Vendor.ViewDashboard,
                Auth.ChangePassword,
                Auth.ViewOwnProfile,
                Auth.EditOwnProfile,
                Product.ViewList,
                Product.View,
                Product.Create,
                Product.Edit
            },
            Roles.Admin => GetAll(),
            Roles.SuperAdmin => GetAll(),
            _ => new List<string>()
        };
    }
}

public static class Policies
{
    public static class Customer
    {
        public const string ViewProfile = "CustomerViewProfile";
        public const string EditProfile = "CustomerEditProfile";
        public const string ViewList = "CustomerViewList";
        public const string Delete = "CustomerDelete";
    }

    public static class Vendor
    {
        public const string ViewProfile = "VendorViewProfile";
        public const string EditProfile = "VendorEditProfile";
        public const string ViewDashboard = "VendorViewDashboard";
        public const string ViewList = "VendorViewList";
        public const string Create = "VendorCreate";
        public const string Delete = "VendorDelete";
    }

    public static class Admin
    {
        public const string ViewProfile = "AdminViewProfile";
        public const string EditProfile = "AdminEditProfile";
        public const string ViewList = "AdminViewList";
        public const string Create = "AdminCreate";
        public const string Delete = "AdminDelete";
        public const string ManageUsers = "AdminManageUsers";
        public const string ManageRoles = "AdminManageRoles";
        public const string ManageClaims = "AdminManageClaims";
    }

    public static class Auth
    {
        public const string ChangePassword = "AuthChangePassword";
        public const string ViewOwnProfile = "AuthViewOwnProfile";
        public const string EditOwnProfile = "AuthEditOwnProfile";
    }

    public static class Product
    {
        public const string ViewList = "ProductViewList";
        public const string View = "ProductView";
        public const string Create = "ProductCreate";
        public const string Edit = "ProductEdit";
        public const string Delete = "ProductDelete";
    }

    public static class Category
    {
        public const string ViewList = "CategoryViewList";
        public const string View = "CategoryView";
        public const string Create = "CategoryCreate";
        public const string Edit = "CategoryEdit";
        public const string Delete = "CategoryDelete";
    }

    public static class SubCategory
    {
        public const string ViewList = "SubCategoryViewList";
        public const string View = "SubCategoryView";
        public const string Create = "SubCategoryCreate";
        public const string Edit = "SubCategoryEdit";
        public const string Delete = "SubCategoryDelete";
    }

    public static class Brand
    {
        public const string ViewList = "BrandViewList";
        public const string View = "BrandView";
        public const string Create = "BrandCreate";
        public const string Edit = "BrandEdit";
        public const string Delete = "BrandDelete";
    }

    public static class UnitOfMeasure
    {
        public const string ViewList = "UnitOfMeasureViewList";
        public const string View = "UnitOfMeasureView";
        public const string Create = "UnitOfMeasureCreate";
        public const string Edit = "UnitOfMeasureEdit";
        public const string Delete = "UnitOfMeasureDelete";
    }

    public static class Warranty
    {
        public const string ViewList = "WarrantyViewList";
        public const string View = "WarrantyView";
        public const string Create = "WarrantyCreate";
        public const string Edit = "WarrantyEdit";
        public const string Delete = "WarrantyDelete";
    }

    public static class VariantAttribute
    {
        public const string ViewList = "VariantAttributeViewList";
        public const string View = "VariantAttributeView";
        public const string Create = "VariantAttributeCreate";
        public const string Edit = "VariantAttributeEdit";
        public const string Delete = "VariantAttributeDelete";
    }

    public static class Account
    {
        public const string ViewList = "AccountViewList";
        public const string View = "AccountView";
        public const string Create = "AccountCreate";
        public const string Edit = "AccountEdit";
        public const string Delete = "AccountDelete";
    }

    public static class Coupon
    {
        public const string ViewList = "CouponViewList";
        public const string View = "CouponView";
        public const string Create = "CouponCreate";
        public const string Edit = "CouponEdit";
        public const string Delete = "CouponDelete";
    }

    public static class Discount
    {
        public const string ViewList = "DiscountViewList";
        public const string View = "DiscountView";
        public const string Create = "DiscountCreate";
        public const string Edit = "DiscountEdit";
        public const string Delete = "DiscountDelete";
    }

    public static class GiftCard
    {
        public const string ViewList = "GiftCardViewList";
        public const string View = "GiftCardView";
        public const string Create = "GiftCardCreate";
        public const string Edit = "GiftCardEdit";
        public const string Delete = "GiftCardDelete";
    }
}
